{"cells":[{"cell_type":"markdown","metadata":{"id":"gQq_-NDM_G89"},"source":["# Errors and Exceptions"]},{"cell_type":"markdown","metadata":{"id":"oUXsB3Zk_G9C"},"source":["![error.gif](attachment:error.gif)"]},{"cell_type":"markdown","metadata":{"id":"DIL9Aii6_G9C"},"source":["If you can hear that image then you are in the right place :)\n","\n","It is always helpful when someone else is indicating what is wrong on our actions in code that's why error messages are important to be read carefully and be taken into consideration seriously.\n","\n","Python has [built-in](https://www.programiz.com/python-programming/exceptions) exceptions to raise and indicate errors in the code. But what is the actual difference between syntax errors and exceptions?\n","\n","**Syntax error** is a mistake in the use of the Python language, and is analogous to spelling or grammar mistakes in a language like English. Such errors are preventing the program to be executed thus it stops.\n","\n","\n","Example:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"fzmChhpD_G9D","colab":{"base_uri":"https://localhost:8080/","height":106},"executionInfo":{"status":"error","timestamp":1709902778365,"user_tz":-60,"elapsed":6,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"2cdcfc2d-8a20-454b-da10-9126b023b2c4"},"outputs":[{"output_type":"error","ename":"SyntaxError","evalue":"expected ':' (<ipython-input-1-f142a8bbbd5f>, line 5)","traceback":["\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-f142a8bbbd5f>\"\u001b[0;36m, line \u001b[0;32m5\u001b[0m\n\u001b[0;31m    if(age > 18)\u001b[0m\n\u001b[0m                ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m expected ':'\n"]}],"source":["# initialize the amount variable\n","age = 30\n","\n","# check if the user is eligible to drink\n","if(age > 18)\n","print(\"You are allowed to drink\")"]},{"cell_type":"markdown","metadata":{"id":"56Joil6l_G9F"},"source":["Error messages provide information about the line that caused the error by tracing back to the function calls that lead to this instruction. The line numbers of the function calls are displayed in the error message to enable quick correction of the code. The parser repeats the offending line and displays a little ‘arrow’ pointing at the earliest point in the line where the error was detected."]},{"cell_type":"markdown","metadata":{"id":"L83YwVXj_G9H"},"source":["**Exception** is when there is nothing wrong in syntax but something is wrong and it changes the flow of the program. Exceptions have a lot of types but below we will see only **3 types** : *ZeroDivisionError, TypeError* and *NameError* each one with its own meaning. Let's see some examples in order to get the differences better.\n","\n","Example 1 :"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"__-tLO13_G9H","colab":{"base_uri":"https://localhost:8080/","height":228},"executionInfo":{"status":"error","timestamp":1709903205873,"user_tz":-60,"elapsed":3017,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"60cc7cd4-b8e0-4448-933f-d05af5722e9a"},"outputs":[{"name":"stdout","output_type":"stream","text":["Enter integer: 6\n"]},{"output_type":"error","ename":"ZeroDivisionError","evalue":"division by zero","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-5777c5fbb630>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Enter integer: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mnum\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0minverse\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minverse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}],"source":["a = input(\"Enter integer: \")\n","num = int(a)\n","inverse = 1 / 0\n","print(num, inverse)\n","\n","## ATTENTION! What happens if the user enters a null or non-numeric value? The program will stop and raise an error as shown below."]},{"cell_type":"markdown","metadata":{"id":"6Ta7BbzU_G9I"},"source":["Example 2:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rtuSS_En_G9I","colab":{"base_uri":"https://localhost:8080/","height":141},"executionInfo":{"status":"error","timestamp":1709903087162,"user_tz":-60,"elapsed":241,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"a1db937f-3dc3-489f-a3be-658cf972d89e"},"outputs":[{"output_type":"error","ename":"NameError","evalue":"name 'y' is not defined","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-5-0662eb7d97fd>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mx\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m5\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0my\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mNameError\u001b[0m: name 'y' is not defined"]}],"source":["x = 5 + y"]},{"cell_type":"markdown","metadata":{"id":"ME83npcH_G9J"},"source":["Example 3:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lGI4cVrl_G9J","colab":{"base_uri":"https://localhost:8080/","height":141},"executionInfo":{"status":"error","timestamp":1709903111271,"user_tz":-60,"elapsed":289,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"78d89746-e021-4a1d-a705-d80f1b3af42b"},"outputs":[{"output_type":"error","ename":"TypeError","evalue":"can only concatenate str (not \"int\") to str","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-7-379e421cfaff>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'2'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mTypeError\u001b[0m: can only concatenate str (not \"int\") to str"]}],"source":["a = '2' + 5"]},{"cell_type":"markdown","metadata":{"id":"NOMgXAFN_G9K"},"source":["Can you spot the differences?\n","\n","Interesting right? You can now understand , more or less, what is the problem with your code. Hourayyyy!\n","\n","Let's now try to see how we can handle such exceptions and make our life easier.\n"]},{"cell_type":"markdown","metadata":{"id":"EX1H5Adm_G9K"},"source":["# Exception handling"]},{"cell_type":"markdown","metadata":{"id":"FP9uimBi_G9L"},"source":["## Try and except\n","\n","**Try** and **except** statements are used in order to catch and eventually handle exceptions in Python. The commands that may cause issues are written inside the try clause and the commands that will handle the exceptionare written inside except clause.\n","\n","\n","As we saw above, if you try to divide with 0, you will get the ZeroDivisionError, so how can we handle that?"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"09anax1x_G9L","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903251587,"user_tz":-60,"elapsed":208,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"2804e528-583d-499d-fee6-f2ff2b6968bb"},"outputs":[{"output_type":"stream","name":"stdout","text":["Trying to divide with zero\n","Finished\n"]}],"source":["try:\n","    3 / 0\n","except:\n","    print(\"Trying to divide with zero\")\n","\n","print(\"Finished\")"]},{"cell_type":"markdown","metadata":{"id":"mNKcUAg0_G9M"},"source":["It worked!!! We didn't get any errors, but instead we got a message which dedicates what is wrong and we can continue our execution as we want.\n","\n","So, how this is actually working:\n","- First, whatever is in the try clause is being executed.\n","- If what is written there, has no issues, then the except clause is been ignored and the entire try statement is completed.\n","- If something is wrong, then the except clause is being executed\n"]},{"cell_type":"markdown","metadata":{"id":"mwZtqJrh_G9M"},"source":["One try clause can have multiple except clauses to specify handlers for different exceptions.\n","\n","Now, we will try to avoid the exceptions mentioned above by rewriting our code as follows.   \n","We have the ability to capture the type of exception."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hWOYpgah_G9M","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903434256,"user_tz":-60,"elapsed":218,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"a66c8e42-ff1f-4665-8811-dfec49a42021"},"outputs":[{"output_type":"stream","name":"stdout","text":["Error: You can't divide by zero\n"]}],"source":["try:\n","    3 / 0\n","except ZeroDivisionError:\n","    print(\"Error: You can't divide by zero\")\n","except ValueError:\n","    print(\"Error: Non-numeric value\")\n","except BaseException:\n","    print(\"Error: there is a problem\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"hvQ_fowF_G9N","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903516064,"user_tz":-60,"elapsed":224,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"68adc190-e46b-4ba7-a037-07b2b08f463f"},"outputs":[{"output_type":"stream","name":"stdout","text":["Error: there is a problem\n"]}],"source":["try:\n","    3 / ko\n","except ZeroDivisionError:\n","    print(\"Error: You can't divide by zero\")\n","except ValueError:\n","    print(\"Error: Non-numeric value\")\n","except BaseException:\n","    print(\"Error: there is a problem\")"]},{"cell_type":"markdown","metadata":{"id":"pyvcdOjS_G9O"},"source":["You can also except multiple exception as one. If you need to know which one was raised, you can save the exception's error message if you want to print it."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Mn4WvIDE_G9O","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903474541,"user_tz":-60,"elapsed":301,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"a7c5dd21-eaea-46be-bbd5-7aa9dc193c27"},"outputs":[{"output_type":"stream","name":"stdout","text":["Error:  invalid literal for int() with base 10: 'ssss'\n"]}],"source":["try:\n","    3 / int(\"ssss\")\n","except (ZeroDivisionError, ValueError, BaseException) as ex:\n","    print(\"Error: \", ex)"]},{"cell_type":"markdown","metadata":{"id":"9ipxuUJ3_G9O"},"source":["### Finally and else\n","\n","As alternatives at the end of our exception handling we can use two other reserved keywords :  `finally` and `else`.  \n","`else` is executed when no other exception is being detected and `finally` is a block that is executed after all other blocks have been executed, regardless of whether there was an exception or not, and **even if the program crashes**."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"rQNExfLh_G9P","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903605182,"user_tz":-60,"elapsed":296,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"0060b5d3-ad18-4b9b-adf3-bc5f0c3679aa"},"outputs":[{"output_type":"stream","name":"stdout","text":["Everything is ok\n"]}],"source":["try:\n","    3 / 3\n","except ZeroDivisionError:\n","    print(\"Error: You can't divide by zero\")\n","except ValueError:\n","    print(\"Error: Non-numeric value\")\n","except BaseException:\n","    print(\"Error: there is a problem\")\n","else:\n","    print(\"Everything is ok\")"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"zSDlDIPN_G9P","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1709903615322,"user_tz":-60,"elapsed":214,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"c1049a41-b47d-4435-8d1d-e852306422d3"},"outputs":[{"output_type":"stream","name":"stdout","text":["Error: You can't divide by zero\n","I'll do executing in the end no matter what..\n"]}],"source":["try:\n","    3 / 0\n","except ZeroDivisionError:\n","    print(\"Error: You can't divide by zero\")\n","except ValueError:\n","    print(\"Error: Non-numeric value\")\n","except BaseException:\n","    print(\"Error: there is a problem\")\n","finally:\n","    print(\"I'll do executing in the end no matter what..\")"]},{"cell_type":"markdown","metadata":{"id":"0Vxco-NE_G9Q"},"source":["### Raise"]},{"cell_type":"markdown","metadata":{"id":"dIfFZN1c_G9Q"},"source":["Sometimes it is fun to make mistakes, same when you are coding. Therefore, we have the statement `raise` which makes it possible to trigger exceptions by ourselves."]},{"cell_type":"markdown","metadata":{"id":"DKjwRcS7_G9R"},"source":["`raise` is a Python statement that can trigger any `Exception`. This means that an error is explicitly triggered."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"AWK4gXlM_G9R","colab":{"base_uri":"https://localhost:8080/","height":297},"executionInfo":{"status":"error","timestamp":1709903679529,"user_tz":-60,"elapsed":240,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"85ba4bad-67b3-4024-95cf-74b1f8f56bb0"},"outputs":[{"output_type":"error","ename":"ZeroDivisionError","evalue":"","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-17-5b1616062354>\u001b[0m in \u001b[0;36m<cell line: 8>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0mdivision\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m<ipython-input-17-5b1616062354>\u001b[0m in \u001b[0;36mdivision\u001b[0;34m(num, div)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mdivision\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdiv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mdiv\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mZeroDivisionError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mnum\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0mdiv\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: "]}],"source":["def division(num, div):\n","    if div == 0:\n","        raise ZeroDivisionError()\n","    else:\n","        return num / div\n","\n","\n","division(5, 0)"]},{"cell_type":"markdown","metadata":{"id":"r3YzEbUD_G9R"},"source":["We can agree that the `division()` function is completely useless! This was just to showcase the purpose of `raise`.\n","\n","### Creating an exception\n","As you can imagine, we can create our own exceptions.\n","Just create a class that will inherit from the `Exception` class.\n","\n","*Note that it is good practice to add docstring to document the exception's purpose.*"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"n_9sihvG_G9S"},"outputs":[],"source":["class MyError(Exception):\n","    \"\"\"Raised when demonstating exception handling.\"\"\"\n","\n","    pass"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Lc96VZN7_G9S","colab":{"base_uri":"https://localhost:8080/","height":141},"executionInfo":{"status":"error","timestamp":1709903725768,"user_tz":-60,"elapsed":212,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"bed97614-e24a-4d5f-db2a-ee07e54ba340"},"outputs":[{"output_type":"error","ename":"MyError","evalue":"Hello","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mMyError\u001b[0m                                   Traceback (most recent call last)","\u001b[0;32m<ipython-input-19-e6e7a22c0cb6>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mraise\u001b[0m \u001b[0mMyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Hello\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mMyError\u001b[0m: Hello"]}],"source":["raise MyError(\"Hello\")"]},{"cell_type":"markdown","metadata":{"id":"FnTFrIJl_G9S"},"source":["### Wrap it together\n","\n","It is common to add a try/except statement around function.\n","\n","A simple example here:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"DYsde5N1_G9T"},"outputs":[],"source":["def division(num, div):\n","    if div == 0:\n","        raise ZeroDivisionError(\"You can't input 0 as a second argument!\")\n","    else:\n","        return num / div\n","\n","try:\n","    division(5, 4)\n","except ZeroDivisionError as ex:\n","    print(\"Can't compute result: \", ex)"]},{"cell_type":"markdown","metadata":{"id":"DmzjSeVe_G9T"},"source":["### Final warning\n","\n","Try/except are great! But as all good things, you should not abuse it.\n","\n","The catch is to try/except everything without scepcific exception.\n","You risk to end up not seeing errors you does not expect without understanding what is wrong.\n","\n","To avoid that, we recommend always to add a print statement when you except something."]},{"cell_type":"markdown","metadata":{"id":"bvY-uour_G9T"},"source":["## Explore further:\n","- https://www.w3schools.com/python/python_try_except.asp\n","- https://www.tutorialspoint.com/python/python_exceptions.htm\n","- https://realpython.com/python-exceptions/"]},{"cell_type":"markdown","metadata":{"id":"2bKbOaYk_G9U"},"source":["## Your turn\n","\n","Based on what you discovered today and on some external research, write a Python code for each of those problems **by using exceptions**:\n","\n","1. Ask the user to input two numbers and perform division. Use a try-except block to handle the ZeroDivisionError exception.\n","2. Create a list of integers. Ask the user to input an index to access an element from the list. Use a try-except block to handle the IndexError exception.\n","3. Ask the user to input a number. Convert the user input to an integer or float using int() or float(). Use a try-except block to handle the ValueError exception.\n","4. Create a custom exception called \"InvalidAgeException\". Write a function that takes an age as input and raises this exception if the age is negative or greater than 150.\n","5. Create a function that takes two numbers as input and performs division. If the divisor is zero, raise a custom exception called \"InvalidDivisorException\". Use a try-except block to handle this exception."]},{"cell_type":"code","source":["#1 Ask the user to input two numbers and perform division. Use a try-except block to handle the ZeroDivisionError exception.\n","\n","def division(a,b):\n","  if b == 0:\n","    raise ZeroDivisionError(\"0 ne peut pas être dénominateur!\")\n","  else:\n","    return print(a/b)\n","\n","numerateur= int(input(\"Entrez un numérateur: \"))\n","denominateur= int(input(\"Entrez un dénominateur: \"))\n","\n","try:\n","   division(numerateur, denominateur)\n","except ZeroDivisionError as ex:\n","  print(\"Error:\", ex)\n","\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"LyeVhG2tQ7r1","executionInfo":{"status":"ok","timestamp":1709905442440,"user_tz":-60,"elapsed":4767,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"ea001d0d-932b-43ea-9690-64cef2e08864"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Entrez un numérateur: 10\n","Entrez un dénominateur: 0\n","Error: 0 ne peut pas être dénominateur!\n"]}]},{"cell_type":"code","source":["#2 Create a list of integers. Ask the user to input an index to access an element from the list. Use a try-except block to handle the IndexError exception.\n","\n","import random\n","\n","numlist=[]\n","while len(numlist) <=8:\n","  numlist.append(random.randint(0,99))\n","print(numlist)\n","\n","def index_access(my_list):\n","  try:\n","    index= int(input(\"Index: \"))\n","    element= my_list[index]\n","    print(\"Le nombre à l'index\", index, \"est:\", element)\n","  except IndexError:\n","    print(\"Index indisponible\")\n","\n","index_access(numlist)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"zeBEsfi7Ton5","executionInfo":{"status":"ok","timestamp":1710251050052,"user_tz":-60,"elapsed":2211,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"d0af152b-38b8-48b0-9459-1a9042b02531"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[83, 18, 25, 9, 46, 96, 90, 58, 18]\n","Index: 0\n","Le nombre à l'index 0 est: 83\n"]}]},{"cell_type":"code","source":["#3 Ask the user to input a number. Convert the user input to an integer or float using int() or float(). Use a try-except block to handle the ValueError exception.\n","\n","x = input(\"Nombre: \")\n","\n","def convert(x):\n","\n","  x=int(x)\n","  return x\n","\n","try:\n","  convert(x)\n","except ValueError as ex:\n","   x = float(x)\n","   print(x)\n","\n","\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Fk6NOyqordcq","executionInfo":{"status":"ok","timestamp":1710249054869,"user_tz":-60,"elapsed":6164,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"3b971faa-fc8c-4408-a013-d7bad695e422"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Nombre: 5.5\n","5.5\n"]}]},{"cell_type":"code","source":["#4 Create a custom exception called \"InvalidAgeException\". Write a function that takes an age as input and raises this exception if the age is negative or greater than 150.\n","\n","class InvalidAgeException(Exception):\n","\n","  pass\n","\n","age=int(input(\"Age: \"))\n","\n","if age > 150:\n","  raise InvalidAgeException(\"Age invalide\")\n","\n","elif age < 0:\n","  raise InvalidAgeException(\"Age invalide\")\n","\n","else:\n","  print(\"Vous avez\", age, \"ans\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":228},"id":"7r6Rz8DbrhOY","executionInfo":{"status":"error","timestamp":1710247765215,"user_tz":-60,"elapsed":6252,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"a9e5985c-1b21-4144-a6af-e9ea3cc14cde"},"execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["Age: -2\n"]},{"output_type":"error","ename":"InvalidAgeException","evalue":"Age invalide","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mInvalidAgeException\u001b[0m                       Traceback (most recent call last)","\u001b[0;32m<ipython-input-4-8862555faf27>\u001b[0m in \u001b[0;36m<cell line: 9>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32melif\u001b[0m \u001b[0mage\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m   \u001b[0;32mraise\u001b[0m \u001b[0mInvalidAgeException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Age invalide\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mInvalidAgeException\u001b[0m: Age invalide"]}]},{"cell_type":"code","source":["#5 Create a function that takes two numbers as input and performs division. If the divisor is zero, raise a custom exception called \"InvalidDivisorException\". Use a try-except block to handle this exception.\n","\n","class InvalidDivisorException(Exception):\n","  pass\n","\n","def division(num, denum):\n","  if denum == 0:\n","    raise InvalidDivisorException(\"0 ne peut pas être au dénominateur\")\n","\n","  else:\n","    return num/denum\n","\n","a=int(input(\"Numérateur: \"))\n","b=int(input(\"Dénominateur: \"))\n","\n","division(a,b)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":332},"id":"mEW-6UpZrjbY","executionInfo":{"status":"error","timestamp":1710248101220,"user_tz":-60,"elapsed":4670,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"1e13a842-3943-40d0-c148-b284ec941494"},"execution_count":null,"outputs":[{"name":"stdout","output_type":"stream","text":["Numérateur: 10\n","Dénominateur: 0\n"]},{"output_type":"error","ename":"InvalidDivisorException","evalue":"0 ne peut pas être au dénominateur","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mInvalidDivisorException\u001b[0m                   Traceback (most recent call last)","\u001b[0;32m<ipython-input-6-4b353f277ca3>\u001b[0m in \u001b[0;36m<cell line: 16>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Dénominateur: \"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 16\u001b[0;31m \u001b[0mdivision\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;32m<ipython-input-6-4b353f277ca3>\u001b[0m in \u001b[0;36mdivision\u001b[0;34m(num, denum)\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mdivision\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdenum\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mdenum\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0;32mraise\u001b[0m \u001b[0mInvalidDivisorException\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"0 ne peut pas être au dénominateur\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m   \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mInvalidDivisorException\u001b[0m: 0 ne peut pas être au dénominateur"]}]},{"cell_type":"markdown","metadata":{"id":"1N9JV4ld_G9U"},"source":[]},{"cell_type":"code","source":["# Guess the number\n","\n","import random\n","\n","limite_haute = 100\n","limite_basse = 1\n","tentative_max = 10\n","\n","nombre_mystere = random.randint(limite_basse, limite_haute)\n","\n","def get_name():\n","  name = input(\"Nom du Joueur: \")\n","  return name\n","\n","def get_guess():\n","  while True:\n","    try:\n","      guess = int(input(f\"Devinez un nombre entre {limite_basse} et {limite_haute}: \"))\n","      if limite_basse <= guess <= limite_haute:\n","        return guess\n","\n","      else:\n","        print(f\"Veuillez entrer un nombre entre {limite_basse} et {limite_haute}\")\n","\n","    except ValueError:\n","      print(\"Input invalide. Veuillez entrez un nombre valable.\")\n","\n","def validate_guess(guess, nombre_mystere):\n","\n","  if guess == nombre_mystere:\n","    return \"Correct\"\n","\n","  elif guess < nombre_mystere:\n","    return \"Trop bas\"\n","\n","  elif guess > nombre_mystere:\n","    return \"Trop élevé\"\n","\n","def play_game():\n","  tentative = 0\n","  name = get_name()\n","  win = False\n","\n","  while tentative < tentative_max:\n","    tentative += 1\n","    guess = get_guess()\n","    result = validate_guess(guess, nombre_mystere)\n","\n","    if result == \"Correct\":\n","      print(f\"Bravo {name} ! Vous avez trouvé le nombre mystere {nombre_mystere} en {tentative} tentative\")\n","      win = True\n","      break\n","\n","    else:\n","      print(f\"{result}, Réessayez!\")\n","\n","  if not win:\n","      print(f\"Dommage {name}, vous n'avez plus de tentative, le nombre mystère était, {nombre_mystere}\")\n","\n","play_game()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0_py0Ibh6SLS","executionInfo":{"status":"ok","timestamp":1710325855395,"user_tz":-60,"elapsed":39661,"user":{"displayName":"Hicham Ahamdi","userId":"17336652271089384723"}},"outputId":"af3b82f3-eed4-48e6-d527-4018d2a2cc77"},"execution_count":30,"outputs":[{"output_type":"stream","name":"stdout","text":["Nom du Joueur: max\n","Devinez un nombre entre 1 et 100: 15\n","Trop bas, Réessayez!\n","Devinez un nombre entre 1 et 100: 56\n","Trop bas, Réessayez!\n","Devinez un nombre entre 1 et 100: 80\n","Trop bas, Réessayez!\n","Devinez un nombre entre 1 et 100: 90\n","Trop élevé, Réessayez!\n","Devinez un nombre entre 1 et 100: 86\n","Trop bas, Réessayez!\n","Devinez un nombre entre 1 et 100: 88\n","Bravo max ! Vous avez trouvé le nombre mystere 88 en 6 tentative\n"]}]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.5"},"vscode":{"interpreter":{"hash":"65c45682600b822032a055cbf04f0dc5abe5e3c026f7225f43ce560cd84c5728"}},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":0}